poly/schnorr.go

```diff
@@ -0,0 +1,350 @@
+	}
+	if s.partials[ps.Index] != nil {
+		return errors.New(fmt.Sprintf("A Partial Signature has already been added for this index %d", ps.Index))
+	}
+	if err := s.verifyPartialSig(ps); err != nil {
+		return errors.New(fmt.Sprintf("Partial signature to add is not valid : %v", err))
+	}
+	s.partials[ps.Index] = ps
+	return nil
+}
+
+// SchnorrSig  will generate the global schnorr signature
+// By reconstructing the secret that the partial responses contains
+func (s *Schnorr) SchnorrSig() (*SchnorrSig, error) {
+	// automatic verification
+	// TODO : change this into a bool flag or public method ?
+	if err := s.verify(); err != nil {
+		return nil, err
+	}
+
+	pri := PriShares{}
+	pri.Empty(s.info.Suite, s.info.T, s.info.N)
+	for i, ps := range s.partials {
+		pri.SetShare(ps.Index, *s.partials[i].Part)
+	}
+	// lagrange interpolation to compute the gamma
+	gamma := pri.Secret()
+	sig := &SchnorrSig{
+		random:    s.random.Pub,
+		signature: &gamma,
+	}
+	return sig, nil
+}
+
+// VerifySchnorrSig will verify if a given signature is correct regarding the message
+// NOTE: This belongs to the schnorr structs however it can be called at any time you want
+// This check is static, meaning it only needs the longterm shared secret, and the signature to
+// check. Think of the schnorr signature as a black box having two inputs
+//  - a message to be signed + a random secret ==> NewRound
+//  - a message + a signature to check on ==> VerifySchnorrSig
+func (s *Schnorr) VerifySchnorrSig(sig *SchnorrSig, msg []byte) error {
+	// gamma * G
+	left := s.info.Suite.Point().Mul(s.info.Suite.Point().Base(), *sig.signature)
+
+	randomCommit := sig.random.SecretCommit()
+	publicCommit := s.Longterm.Pub.SecretCommit()
+	hash, err := s.hashMessage(msg, randomCommit)
+
+	if err != nil {
+		return err
+	}
+	// RandomSecretCOmmit + H( ...) * LongtermSecretCommit
+	right := s.info.Suite.Point().Add(randomCommit, s.info.Suite.Point().Mul(publicCommit, hash))
+
+	if !left.Equal(right) {
+		return errors.New("Signature could not have been verified against the message")
+	}
+	return nil
+}
+
+//// DECODING / ENCODING /////
+// Decoding :
+// Use the Schnorr struct to produce already initialized struct
+// 	then call Unmarshal on them. Theses functions are in Schnorr
+// so you can not forget to init the struct with the right fields.
+// Encoding is simple as calling MarshalBinary on them
+
+// Wrapper function to help for the decoding
+func (s *Schnorr) EmptyPartialSig() *PartialSchnorrSig {
+	return new(PartialSchnorrSig).Init(s.info.Suite)
+}
+func (s *Schnorr) EmptySchnorrSig() *SchnorrSig {
+	return new(SchnorrSig).Init(s.info)
+}
+
+// Init is needed before decoding a PartialSchnorrSig so it can now about the length of a secret
+func (pss *PartialSchnorrSig) Init(suite abstract.Suite) *PartialSchnorrSig {
+	pss.Suite = suite
+	return pss
+}
+
+// Marshal a PartialSchnorrSig into a packet
+func (pss *PartialSchnorrSig) MarshalBinary() ([]byte, error) {
```

Comment:
Indeed you are right. I didn't know about abstract/encoding.go and just understood a lot better your abstract/ folder. I've been trying to make a little example out of the abstract.Read / Write method to see if they work like I expect them to.... But unfortunatly I can not have a working example.  Would you be willing to see what's wrong in this example ?
```Go
package main

import (
	"bytes"
	"fmt"
	"github.com/dedis/crypto/abstract"
	"github.com/dedis/crypto/edwards"
)

var suite = edwards.NewAES128SHA256Ed25519(true)

type CustomStruct struct {
	Index int32
	//Secret abstract.Secret
	Index2 int32
}

func main() {
	testReadWrite()
}

func (cs *CustomStruct) Equal(cs2 *CustomStruct) bool {
	if cs.Index != cs2.Index {
		fmt.Printf("Index 1 is not equal\n")
		return false
	}
	if cs.Index2 != cs2.Index2 {
		fmt.Printf("Index 2 is not equal\n")
		return false
	}
	//&& cs.Secret.Equal(cs2.Secret)
	return true
}

func testReadWrite() {
	//sec := suite.Secret().Zero()
	cs := CustomStruct{
		Index: 4,
		//Secret: sec
		Index2: -4,
	}
	var buf *bytes.Buffer = new(bytes.Buffer)
	err := abstract.Write(buf, cs, suite)
	if err != nil {
		fmt.Printf("Hoho error Writing : %v\n", err)
	}

	bytesBuffer := buf.Bytes()

	buf2 := bytes.NewBuffer(bytesBuffer)
	cs2 := CustomStruct{}
	err = abstract.Read(buf2, cs2, suite)
	if err != nil {
		fmt.Printf("Hoho error Reading : %v\n", err)
	}

	if cs.Equal(&cs2) {
		fmt.Printf("YOUPI\n")
	} else {
		fmt.Printf("NOOOOAAAAA\n")
	}
}
```

I've tried with simple "int", then "int32", then removing the "secret" but nothing works ><. 
I am willing to use the abstract.Read / Write if I understand them correctly though. In the meantime, I think the priority like we said is in the comparison baseline.